library/Freckle/App/Memcached/Client.hs

```diff
@@ -95,6 +96,22 @@ set k v expiration = traced $ with $ \case
               ]
         }
 
+-- | Delete a key
+delete
+  :: (MonadUnliftIO m, MonadTracer m, MonadReader env m, HasMemcachedClient env)
+  => CacheKey
+  -> m ()
+delete k = traced $ with $ \case
+  MemcachedClient mc -> void $ liftIO $ Memcache.delete mc (fromCacheKey k) 0
```

Comment:
**Big TODO**: this is passing `0` for the version argument (you have to say which version you want to delete), which doesn't work. I think there is no option other than to do a `get`-`delete` here, but at least we could hide it within this function.

Reply:
> which doesn't work

Wait, what? That's what we're doing in the session library https://github.com/freckle/yesod-session-persist/blob/757f7a7355d0afdab4b8ab39e335daafee7191d3/internal/Yesod/Session/Memcache/Storage.hs#L57

Reply:
Oh nice. You read the docs more thoroughly than I did.

https://github.com/freckle/yesod-session-persist/blob/757f7a7355d0afdab4b8ab39e335daafee7191d3/internal/Yesod/Session/Memcache/Storage.hs#L101-L112

This should be fine then!

Reply:
Oh, do you mean "doesn't work" as in it doesn't have the appropriate behavior in this use case? Because you actually want the check-and-set here to make sure you're evicting the old value?

Reply:
Hiding a get-and-delete within this function, in this case, defeats the purpose.

I'm confused why you need to evict at all. Why not just let the new fetch overwrite the old value?

Reply:
Nope, I was just mistaken. I didn't know `0` had specially meaning and I though this code would only actually delete something if it had the literal version 0, which I expected to not be right. Happy to find out it works differently.

Reply:
> I'm confused why you need to evict at all. Why not just let the new fetch overwrite the old value?

This is a good point. I'll admit that I didn't actually have a reason and the code just ended up that way.

But thinking now, there is one scenario where it matters:

1. GET => 200 Cache-Control max-age=5 ETag=x
2. 6 seconds pass, our cache is stale, we'll use ETag and hope for a 304
3. GET If-Non-Match: x => 200 Cache-Control no-store ETag=y

At step (3) we need to evict, but should not overwrite with anything new.

The test suite sets up this scenario, but not because I thought about it as A Thing To Handle, but because it was the only way to observe eviction at all.

---

Things would be a lot simpler if I didn't care about this case. It drops one whole operation out of the `HttpCache` record and simplifies the stair-`case`, so I'm wondering... should I not care?